README
Acme-MobileTradeFun

I recently found a site called MobileTradeFun (or MobaTreFun for short), where
people can search the value of cards from mobile games, such as idolmaster.

http://mobile-trade.jp

This site seems to contain the "official" card images as they show up on the
mobile devices in Japan.  This module basically lets you specify a game, and
scrapes all the high quality cards off the site.

Note that card names are stored in Japanese, as each card contains certain
attributes such as category, card name and rarity.  Without these attributes,
cards cannot be identified uniquely.

The interface of the module is pretty simple.  You construct a hashref of args,
and pass it onto the class method run() which will take the rest.

    use Acme::MobileTradeFun;

    my $args = {
        game        => 'idolmaster',
        parallel    => 10,
    };

    my $foo = Acme::MobileTradeFun->run( $args );
    
Here are the overwritable defaults in the hashref.  Currently 4 games are
supported: bahamut, idolmaster, saintseiya and gangroad.

    game        => '',
    base_url    => 'http://mobile-trade.jp/fun',
    php_script  => 'card.php',
    row         => 100,     # how many cards per page
    page        => 1,       # which page to start from
    output_dir  => '/tmp',  # where to save the images
    debug       => 0,

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Acme::MobileTradeFun

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Acme-MobileTradeFun

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Acme-MobileTradeFun

    CPAN Ratings
        http://cpanratings.perl.org/d/Acme-MobileTradeFun

    Search CPAN
        http://search.cpan.org/dist/Acme-MobileTradeFun/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Satoshi Yagi

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Acme-MobileTradeFun

0.12    2014/3/1
        Refactored parser into strategy pattern.  The upstream generates
        2 kinds of HTMLs and parsing had to be adopted.

0.11    2013/12/30
        Updated comments per Uncle Bob's style

0.10    2013/12/23
        Updated Mojo::DOM::attr from attrs per API change

0.09    2013/2/23
        Added ability to only fetch new cards, instead of everything.  Removed
        spaces from card names in case they creep in.

0.08    2013/1/29
        Added basic tests for the constructor.

0.07    2012/12/3
        Improved the parsing of categories to work with mixed versions of PHP
        codes at mobile-trade.jp.

0.06    2012/11/14
        Cleaned up the modules used and fixed PREREQ_PM.

0.05    2012/11/14
        Switched the fetching mechanism from forking to event-driven.

0.04    2012/11/5
        Re-added a check to see if game is supplied.

0.03    2012/11/5
        Changed scraping mechanism from regex to Mojo::DOM.  Took away card
        offset and game-to-IP mapping as I no longer look for an IP address in
        the links.

0.02    2012/11/2
        Fixing POD.
        
0.01    2012/11/2
        Initial version.

Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Acme::MobileTradeFun',
    AUTHOR           => q{Satoshi Yagi <satoshi.yagi@yahoo.com>},
    VERSION_FROM     => 'lib/Acme/MobileTradeFun.pm',
    ABSTRACT_FROM    => 'lib/Acme/MobileTradeFun.pm',
    LICENSE          => 'Artistic_2_0',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'LWP::Simple'           => 0,
        'Log::Log4perl'         => 0,
        'URI::Encode'           => 0,
        'AnyEvent'              => 0,
        'AnyEvent::HTTP'        => 0,
        'Mojo::DOM'             => 0,
        'Encode'                => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Acme-MobileTradeFun-*' },
);

lib/Acme/MobileTradeFun/NewParser.pm
package Acme::MobileTradeFun::NewParser;

use strict;
use warnings;
use Mojo::DOM;
use Log::Log4perl qw/:easy/;
use URI::Encode qw/uri_decode/;
use Acme::MobileTradeFun::Utils;
use Data::Dumper;

sub new {
    my ( $class, $args ) = @_;
    my $self = {};
    $self->{ utils } = Acme::MobileTradeFun::Utils->new();
    return bless $self, $class;
}

=head2 parse_data

    parses HTML, populates $self->{ data } hash

=cut

sub parse_data {
    my ( $self, $html, $all_cards ) = @_;

    my $card_found = 0;
    my $dom = Mojo::DOM->new( $html );
    my @new_cards;

    for my $card ( $dom->find( 'section.card' )->each ) {
        my $category = ( $card->find( 'span' )->each )[0]->text;
        my @links = $card->find( 'a' )->each;
        
        my $name = $links[1]->text;
        my $rarity = $links[2]->text;
        $rarity =~ s/\(//;
        $rarity =~ s/\)//;
        my $link = $links[3]->attr( 'href' );

        unless ( $category && $name && $rarity && $link ) {
            my $message = "Something is missing:";
            $message .= " $category" if ( $category );
            $message .= " $name" if ( $name );
            $message .= " $rarity" if ( $rarity );
            $message .= " $link" if ( $link );
            DEBUG $message;
            next;
        }

        my $url = uri_decode( $link );
        my $card_name = "[$category]$name($rarity).jpg";
        $card_name =~ s/\s+//g; # sometimes bunch of spaces creep in
        $card_found++;

        if ( $self->{ utils }->is_new_card( $card_name, $all_cards ) ) {
            my $elem = { name => $card_name, url => $url };
            push @new_cards, $elem;
        }
    }
    return ( $card_found, @new_cards );
}

1;

lib/Acme/MobileTradeFun/OldParser.pm
package Acme::MobileTradeFun::OldParser;

use strict;
use warnings;
use Mojo::DOM;
use Log::Log4perl qw/:easy/;
use URI::Encode qw/uri_decode/;
use Acme::MobileTradeFun::Utils;

sub new {
    my ( $class, $args ) = @_;
    my $self = {};
    $self->{ utils } = Acme::MobileTradeFun::Utils->new();
    return bless $self, $class;
}

=head2 parse_data

    parses HTML, populates $self->{ data } hash

=cut

sub parse_data {
    my ( $self, $html, $all_cards ) = @_;

    my $card_found = 0;
    my $dom = Mojo::DOM->new( $html );
    my @new_cards;
    
    for my $table ( $dom->find( 'table.card_search_result_table' )->each ) {
        my $category;

        # for each span in the table, I am looking for style attribute with
        # color in it -- this should be the tag with category in it
        for my $span ( $table->find( 'span' )->each ) {
            my $color = $span->{ style };
            $category = $span->text if ( $color && $color =~ /color/ );
        }

        my @links = $table->find( 'a' )->each;
        my $name = $links[0]->text;
        my $rarity = $links[1]->text;
        my $link = $links[2]->attr( 'href' );
        
        unless ( $category && $name && $rarity && $link ) {
            my $message = "Something is missing:";
            $message .= " $category" if ( $category );
            $message .= " $name" if ( $name );
            $message .= " $rarity" if ( $rarity );
            $message .= " $link" if ( $link );
            DEBUG $message;
            next;
        }
        
        my $url = uri_decode( $link );
        my $card_name = "[$category]$name($rarity).jpg";
        $card_name =~ s/\s+//g; # sometimes bunch of spaces creep in
        $card_found++;

        if ( $self->{ utils }->is_new_card( $card_name, $all_cards ) ) {
            my $elem = { name => $card_name, url => $url };
            push @new_cards, $elem;
        }
    }
    return ( $card_found, @new_cards );
}

1;

lib/Acme/MobileTradeFun.pm
package Acme::MobileTradeFun;

use 5.006;
use strict;
use warnings;
use Data::Dumper;
use Carp;
use LWP::Simple;
use File::Path qw/make_path/;
use Log::Log4perl qw/:easy/;
use AnyEvent;
use AnyEvent::HTTP;
use Encode;
use Acme::MobileTradeFun::NewParser;
use Acme::MobileTradeFun::OldParser;

=head1 NAME

Acme::MobileTradeFun - Needlessly OO module to scrape card images off of the
MobileTradeFun site

=head1 VERSION

Version 0.12

=cut

our $VERSION = '0.12';


=head1 SYNOPSIS

I recently found a site called MobileTradeFun (or MobaTreFun for short), where
people can search the value of cards from mobile games, such as idolmaster.

http://mobile-trade.jp

This site seems to contain the "official" card images as they show up on the
mobile devices in Japan.  This module basically lets you specify a game, and
scrapes all the high quality cards off of the site.

Note that card names are stored in Japanese, as each card contains certain
attributes such as category, card name and rarity.  Without these attributes,
cards cannot be identified uniquely.

The interface of the module is pretty simple.  You construct a hashref of args,
and pass it onto the class method run() which will take care of the rest.

    use Acme::MobileTradeFun;

    my $args = {
        game        => 'idolmaster',
    };

    my $foo = Acme::MobileTradeFun->run( $args );
    
Here are the overwritable defaults in the hashref.  Currently 4 games are
supported: bahamut, idolmaster, saintseiya and gangroad.

    game        => '',
    base_url    => 'http://mobile-trade.jp',
    php_script  => 'card.php',
    row         => 100,     # how many cards per page
    page        => 1,       # which page to start from
    output_dir  => '/tmp',  # where to save the images
    debug       => 0,


=head1 SUBROUTINES/METHODS

=head2 new

    The constructor

=cut

sub new {
    my ( $class, $args ) = @_;
    my $self = {};
    bless $self, $class;
    $self->init( $args );
    return $self;
}

=head2 init

    Sets up default args, overrides, validates args etc

=cut

sub init {
    my ( $self, $args ) = @_;
    
    croak "not a hashref" if ( $args && ref( $args ) ne "HASH" );

    $self->{ opts } = {
        game        => '',
        base_url    => 'http://mobile-trade.jp',
        php_script  => 'card.php',
        row         => 100,
        page        => 1,
        output_dir  => '/tmp',
        debug       => 0,
    };

    %{ $self->{ opts } } = ( %{ $self->{ opts } }, %{ $args } ) if ( $args );
    
    my $game = $self->{ opts }->{ game };
    croak "game not specified" unless( $game );
    
    $self->{ parser } = $self->_parser_strategy( $game );
    
    Log::Log4perl->easy_init($DEBUG) if ( $self->{ opts }->{ debug } );
    my $out_dir = $self->{ opts }->{ output_dir } . "/" . $self->{ opts }->{ game };
    make_path( $out_dir ) unless( -d $out_dir );
    $self->{ opts }->{ output_dir } = $out_dir; # appending game name as subdir
}

sub _parser_strategy {
    my ( $self, $game ) = @_;
    
    my @newparser_games = qw/
        mobamasu
    /;
    
    my @oldparser_games = qw/
        saintseiya
    /;
    
    return Acme::MobileTradeFun::NewParser->new() if ( grep( /$game/, @newparser_games ) );
    
    if ( grep( /$game/, @oldparser_games ) ) {
        $self->{ opts }->{ base_url } .= "/fun";
        return Acme::MobileTradeFun::OldParser->new();
    }
    croak "could not find parser for $game";
}

=head2 run

    The driver method.  Calls new then other methods to drive.

=cut

sub run {
    my ( $class, $args ) = @_;
    my $self = $class->new( $args );
    $self->load_existing_cards();
    $self->fetch_cards();
    return $self;
}

=head2 load_existing_cards

    Loads existing cards into $self->{ cards } array.

=cut

sub load_existing_cards {
    my $self = shift;

    my $dir = $self->{ opts }->{ output_dir };
    opendir( my $dh, $dir );
    
    while( my $file = readdir( $dh ) ) {
        # need to decode UTF-8 to internal format
        $file = decode( 'UTF-8', $file );
        push @{ $self->{ cards } }, $file if ( $file =~ /\.jpg$/ );
    }
}

=head2 fetch_cards

    Wrapper method to initiate the fetching of cards

=cut

sub fetch_cards {
    my $self = shift;
    
    my $page        = $self->{ opts }->{ page };
    my $base_url    = $self->{ opts }->{ base_url };
    my $script      = $self->{ opts }->{ php_script };
    my $row         = $self->{ opts }->{ row };
    my $game        = $self->{ opts }->{ game };
    my $dir         = $self->{ opts }->{ output_dir };
    my $parser      = $self->{ parser };

    my @newcards;

    while( 1 ) {
        my $url = "$base_url/$game/$script?row=$row&page=$page";
        my $data = get( $url );
        my ( $cards, @new ) = $parser->parse_data( $data, $self->{ cards } );
        push @newcards, @new;
        DEBUG "scraping $url.  $cards found.";
        last unless ( $cards );
        $page++;
    }

    unless( @newcards ) {
        DEBUG "no card to fetch, exiting";
        return;
    }
    
    my $numcards = @newcards;
    DEBUG "we have $numcards new cards.  fetch starting...";

    my $cv = AE::cv {
        DEBUG "fetched all cards!";
    };

    for my $key ( @newcards ) {
        $cv->begin;
        my $url = $key->{ url };
        http_get $url, sub {
            my ( $data, $hdr ) = @_;
            if ( $hdr->{ Status } =~ /^2/ ) {
                my $file = "$dir/$key->{ name }";
                open( my $fh, ">", $file ) or croak "couldn't open $file: $!";
                print $fh $data;
                close $fh or croak "couldn't close $file: $!";
            }
            else {
                DEBUG "something went wrong with $url";
            }
            $cv->end;
        }
    }
    $cv->recv;
}

=head1 AUTHOR

Satoshi Yagi, C<< <satoshi.yagi at yahoo.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-app-mobiletradefun at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Acme-MobileTradeFun>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Acme::MobileTradeFun


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Acme-MobileTradeFun>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Acme-MobileTradeFun>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Acme-MobileTradeFun>

=item * Search CPAN

L<http://search.cpan.org/dist/Acme-MobileTradeFun/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012-2013 Satoshi Yagi.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Acme::MobileTradeFun
